### ISSUE TITLE
failed to retrieve memory content when domain under cloudflare

### ISSUE THREAD

**Issue by BeyondIM** [negrutiu/nsis-nsxfer#3]
The code has been working perfectly before, but recently when I migrate my domain name to cloudflare, the code couldn't get memory content, It show Remote-Content: "".Is there any way to solve this problem?

Thank you very much.

              You can retrieve the memory content like this:
```nsis
NSxfer::Request /URL "https://httpbin.org/get?param1=1&param2=2" /LOCAL memory /END
Pop $0		; Transfer ID
NSxfer::Wait /ID $0
Pop $1		; Wait for transfer to complete
NSxfer::Query /ID $0 /CONTENT /END
Pop $1		; Remote-content as string truncated to ${NSIS_MAX_STRLEN}. Empty string in case of errors
DetailPrint 'Remote-Content: "$1"'
```

Indeed, I've switched developent to [NScurl](https://github.com/negrutiu/nsis-nscurl).
It's built on top of [libcurl](https://curl.haxx.se/libcurl) which I think it's far superior to the native windows API.
With `NScurl` the same code would look like:
```nsis
NScurl::http GET "https://httpbin.org/get?param1=1&param2=2" "MEMORY" /RETURN "@recvdata@" /END
Pop $0
DetailPrint 'Remote-Content: "$0"'
```

`NScurl` is newer and better. I recommend it.
Thanks for your kind words and enjoy my NSIS plugin(s)...

_Originally posted by @negrutiu in https://github.com/negrutiu/nsis-nsxfer/issues/2#issuecomment-662864599_

**Comment by negrutiu** [comment]
Could you share an URL where this happens?
Are us using NScurl or NSxfer?
OS version?

**Comment by BeyondIM** [comment]
I will provide a detailed report for both plugins and multi os versions tomorrow, tks

**Comment by BeyondIM** [comment]
# test NScurl or NSxfer on domain non under cloudflare

url: https://ylnav.com/test/test.txt

<img width="275" alt="2023-03-13_09-21" src="https://user-images.githubusercontent.com/2209318/224587911-14d667ed-0c20-4995-a550-42fc9ddda293.png">

<img width="861" alt="2023-03-13_09-20_1" src="https://user-images.githubusercontent.com/2209318/224588197-b51a3977-a458-4130-9871-2fa47431782c.png">

<img width="411" alt="2023-03-13_09-20" src="https://user-images.githubusercontent.com/2209318/224588222-0fc3ecb2-1a18-44c2-9fb7-d4cf703ec049.png">


# test NScurl or NSxfer on domain under cloudflare

url: https://portableappk.com/test/test.txt

<img width="280" alt="2023-03-13_09-21_1" src="https://user-images.githubusercontent.com/2209318/224588111-6410feb3-f7d1-4b61-adb8-0771a95db6b1.png">

<img width="889" alt="2023-03-13_09-15" src="https://user-images.githubusercontent.com/2209318/224588290-cbb9c57d-7599-4059-804b-39c7c0ae3f33.png">

test on windows 7
<img width="409" alt="2023-03-13_09-19" src="https://user-images.githubusercontent.com/2209318/224588326-ea602cc0-e32f-4aaa-b4b2-98fcf6c40ec1.png">

test on windows 10
<img width="525" alt="2023-03-13_09-22" src="https://user-images.githubusercontent.com/2209318/224588344-9ce407aa-0bb3-4fda-94e9-dc527a6b73a6.png">

**Comment by negrutiu** [comment]
@BeyondIM, thanks for the detailed report, I was able to repro the problem right away.

The issue was caused by NSxfer refusing to download remote content of **unknown size**.
The Cloudflare server doesn't send the `Content-Length` http header.
Downloads to `memory` needed this value to preallocate the memory buffer. Regular downloads to file are not affected by this.

The issue was fixed in the latest release.
Give it a try and see how it goes.
